File: packages/core-dart/lib/src/address/detect.dart
import 'dart:typed_data';

import 'codes.dart';
import '../util/strkey.dart';

/// Classifies a raw Stellar address string into one of the three Stellar
/// address kinds — classic (`G…`), muxed (`M…`), or contract (`C…`) —
/// without performing any deposit-routing logic.
///
/// This is a low-level structural primitive. Callers that need full
/// deposit-routing decisions (routing ID resolution, memo handling, warning
/// codes) should use `extractRouting` instead of calling [detect] directly.
///
/// ---
///
/// ## Address kinds
///
/// | Prefix | [AddressKind] | Description |
/// |--------|---------------|-------------|
/// | `G…`   | [AddressKind.g] | Classic 56-character Stellar account. Valid destination for classic `Payment` operations. Version byte `0x30`, decoded length 35 bytes. |
/// | `M…`   | [AddressKind.m] | Muxed account — a `G` address with a 64-bit integer ID embedded in a single StrKey string. Used for pooled-account deposit routing without a memo. Version byte `0x60`, decoded length 43 bytes. |
/// | `C…`   | [AddressKind.c] | Soroban smart-contract address. Valid StrKey **but not** a valid destination for classic `Payment` operations. The routing layer above this function surfaces contract destinations as `INVALID_DESTINATION`. Version byte `0x10`, decoded length 35 bytes. |
///
/// ---
///
/// ## Input contract
///
/// - [address] may be **any** [String]. [detect] is **contractually
///   non-throwing**: it catches all internal exceptions and returns `null`
///   instead of propagating them. This guarantee covers empty strings,
///   non-ASCII text, strings of any length, and any other arbitrary input.
/// - The prefix check is **case-insensitive**: both `'g…'` and `'G…'` are
///   recognised as a classic address prefix. The caller does not need to
///   normalise casing before calling [detect].
///
/// ---
///
/// ## Return value
///
/// Returns an [AddressKind] on **success**, or `null` on **any failure**.
///
/// `null` is returned for every one of the following conditions:
///
/// - [address] is empty.
/// - The first character is not `G`, `M`, or `C` (case-insensitive) —
///   **unknown prefix**. Prefix detection runs *before* Base32 decoding and
///   checksum verification, so `"NOTANADDRESS"` returns `null` due to an
///   unknown prefix, not a checksum error.
/// - Base32 decoding fails (malformed alphabet, padding error, etc.).
/// - The decoded byte sequence is shorter than 3 bytes.
/// - The CRC-16 checksum of the payload does not match the two trailing bytes
///   of the decoded data — **invalid checksum** (e.g. a tampered G-length
///   string).
/// - The version byte or decoded length does not match the expected values for
///   the detected prefix — **invalid length / version**.
/// - Any unexpected exception is thrown internally.
///
/// All failure modes collapse to the same `null` sentinel. Structured error
/// codes (`UNKNOWN_PREFIX`, `INVALID_CHECKSUM`, `INVALID_LENGTH`) are
/// surfaced by the routing layer that wraps [detect], not by [detect] itself.
///
/// ---
///
/// ## Spec invariants
///
/// The following invariants hold unconditionally and are validated by the
/// shared `spec/vectors.json` test suite run identically across all three
/// language implementations (TypeScript, Go, Dart):
///
/// - **Prefix fires first.** `UNKNOWN_PREFIX` (i.e. `null`) is returned
///   before any checksum or length check is attempted.
/// - **`C` is structurally valid.** A contract address that passes all
///   structural checks returns [AddressKind.c], not `null`. The
///   `INVALID_DESTINATION` routing error is applied by the layer above.
/// - **uint64 muxed IDs.** Muxed account IDs are unsigned 64-bit integers.
///   Any implementation that coerces them to a 64-bit signed integer or a
///   floating-point `double` will silently corrupt IDs above 2^53. The
///   mandatory 2^53 + 1 canary vector (`id: "9007199254740993"`) in the spec
///   suite catches this class of bug.
/// - **Output is canonical.** The [AddressKind] values are stable enum
///   members; consumers must not compare them by name string.
///
/// ---
///
/// ## Examples
///
/// ```dart
/// // Classic G address — returns AddressKind.g
/// final g = detect('GAYCUYT553C5LHVE2XPW5GMEJT4BXGM7AHMJWLAPZP53KJO7EIQADRSI');
/// assert(g == AddressKind.g);
///
/// // Muxed M address — returns AddressKind.m
/// final m = detect(
///     'MAYCUYT553C5LHVE2XPW5GMEJT4BXGM7AHMJWLAPZP53KJO7EIQACAAAAAAAAAAAAD672');
/// assert(m == AddressKind.m);
///
/// // Contract C address — structurally valid, INVALID_DESTINATION only in routing
/// final c = detect('CDLZFC3SYJYDZT7K67VZ75HPJVIEUVNIXF47ZG2FB2RMQQVU2HHGCYSC');
/// assert(c == AddressKind.c);
///
/// // Empty input — returns null
/// assert(detect('') == null);
///
/// // Unknown prefix — returns null (prefix check fires before checksum)
/// assert(detect('NOTANADDRESS') == null);
///
/// // Lowercase input — prefix check is case-insensitive, still classifies correctly
/// final lower = detect('gaycuyt553c5lhve2xpw5gmejt4bxgm7ahmjwlapzp53kjo7eiqadrsi');
/// assert(lower == AddressKind.g);
///
/// // Tampered checksum — returns null
/// assert(detect('GAYCUYT553C5LHVE2XPW5GMEJT4BXGM7AHMJWLAPZP53KJO7EIQADRSJ') == null);
/// ```
///
/// ---
///
/// ## Spec compliance
///
/// The behaviour of [detect] is fully specified in `spec/vectors.json` at the
/// repository root. All three language implementations produce identical
/// results for every vector in the suite. The spec is the source of truth —
/// if a vector passes in TypeScript, it passes identically in Go and Dart.
///
/// See also:
///
/// - [AddressKind] for the enumeration of the three Stellar address types.
/// - `extractRouting` for full deposit-routing logic built on top of [detect],
///   including memo handling, routing ID resolution, and structured warning
///   codes.
/// - `StrKeyUtil` for the Base32 decoding and CRC-16 checksum primitives used
///   internally by this function.
AddressKind? detect(String address) {
  if (address.isEmpty) return null;

  final prefix = address[0].toUpperCase();
  if (prefix != 'G' && prefix != 'M' && prefix != 'C') return null;

  try {
    final decoded = StrKeyUtil.decodeBase32(address);
    if (decoded.length < 3) return null;

    final payload = decoded.sublist(0, decoded.length - 2);
    final checksum = decoded.sublist(decoded.length - 2);
    final calculated =
        StrKeyUtil.calculateChecksum(Uint8List.fromList(payload));

    if (checksum[0] != (calculated & 0xFF) ||
        checksum[1] != ((calculated >> 8) & 0xFF)) {
      return null;
    }

    // Enforce exact version byte and decoded length for each address kind.
    final versionByte = payload[0];
    switch (prefix) {
      case 'G':
        if (decoded.length != 35 || versionByte != 0x30) return null;
        return AddressKind.g;
      case 'M':
        if (decoded.length != 43 || versionByte != 0x60) return null;
        return AddressKind.m;
      case 'C':
        if (decoded.length != 35 || versionByte != 0x10) return null;
        return AddressKind.c;
      default:
        return null;
    }
  } catch (_) {
    return null;
  }
}

File: packages/core-ts/src/address/detect.ts
import { StrKey } from "@stellar/stellar-sdk";

const BASE32_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";

/**
 * Decodes a Base32-encoded string into binary data.
 * @internal
 */
function decodeBase32(input: string): Uint8Array {
  const s = input.toUpperCase().replace(/=+$/, "");
  const byteCount = Math.floor((s.length * 5) / 8);
  const result = new Uint8Array(byteCount);
  let buffer = 0;
  let bitsLeft = 0;
  let byteIndex = 0;
  for (const ch of s) {
    const value = BASE32_CHARS.indexOf(ch);
    if (value === -1) throw new Error(`Invalid base32 character: ${ch}`);
    buffer = (buffer << 5) | value;
    bitsLeft += 5;
    if (bitsLeft >= 8) {
      if (byteIndex < byteCount) {
        result[byteIndex++] = (buffer >> (bitsLeft - 8)) & 0xff;
      }
      bitsLeft -= 8;
      buffer &= (1 << bitsLeft) - 1;
    }
  }
  return result;
}

/**
 * Computes a 16-bit CRC (CCITT-FALSE) for binary validation.
 * @internal
 */
function crc16(bytes: Uint8Array): number {
  let crc = 0;
  for (const byte of bytes) {
    crc ^= byte << 8;
    for (let i = 0; i < 8; i++) {
      if (crc & 0x8000) {
        crc = (crc << 1) ^ 0x1021;
      } else {
        crc <<= 1;
      }
      crc &= 0xffff;
    }
  }
  return crc;
}

/**
 * Identifies the Stellar address kind from a string input.
 * Supports G (Ed25519), M (Muxed/SEP-23), and C (Contract) addresses.
 * Returns "invalid" if the input does not match a supported format.
 */
export function detect(address: string): "G" | "M" | "C" | "invalid" {
  if (!address) return "invalid";
  const up = address.toUpperCase();

  if (StrKey.isValidEd25519PublicKey(up)) return "G";
  if (StrKey.isValidMed25519PublicKey(up)) return "M";
  if (StrKey.isValidContract(up)) return "C";

  try {
    const prefix = up[0];
    if (prefix === "M") {
      const decoded = decodeBase32(up);
      // M-addresses (SEP-23) payload consists of 1 version byte (0x60),
      // 32-byte pubkey, 8-byte ID, and 2-byte CRC16 checksum.
      if (decoded.length === 43 && decoded[0] === 0x60) {
        const data = decoded.slice(0, decoded.length - 2);
        const checksum =
          decoded[decoded.length - 2] | (decoded[decoded.length - 1] << 8);
        if (crc16(data) === checksum) {
          return "M";
        }
      }
    }
  } catch {
    // Return "invalid" on any decoding failure.
  }

  return "invalid";
}
